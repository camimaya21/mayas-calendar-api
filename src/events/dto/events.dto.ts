import { IsDateString, IsUUID } from 'class-validator';
import { Event } from '@prisma/client';

export class CreateEventDTO {
  title: string;
  description: string;
  @IsDateString()
  startDate: string;
  @IsDateString()
  endDate: string;
}

export class EventDTO {
  @IsUUID()
  id: string;
  title: string;
  description: string;
  @IsDateString()
  startDate: Date;
  @IsDateString()
  endDate: Date;

  public static entityToDTO(event: Event): EventDTO {
    return {
      id: event.id,
      title: event.title,
      description: event.description,
      startDate: event.start_date,
      endDate: event.end_date
    };
  }
}
