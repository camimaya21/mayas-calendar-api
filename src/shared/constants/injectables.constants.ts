export enum REPOSITORIES {
  IEventsRepository = 'IEventsRepository'
}

export enum SERVICES {
  IEventsService = 'IEventsService'
}
