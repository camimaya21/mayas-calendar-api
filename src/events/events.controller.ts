import { Body, Delete, Get, Inject, Param, Post, Put, Query, Controller } from '@nestjs/common';
import { SERVICES } from '../shared/constants/injectables.constants';
import { CreateEventDTO, EventDTO, UpdateEventDTO } from './dto/events.dto';
import { IEventsService, ParamsFindAllEvents, ParamsFindSingleEvent } from './types/events.types';

@Controller('events')
export class EventsController {
  constructor(@Inject(SERVICES.IEventsService) protected readonly eventsService: IEventsService) {}

  @Get('')
  async findAllEvents(@Query() params: ParamsFindAllEvents): Promise<EventDTO[]> {
    return this.eventsService.findAll(params);
  }

  @Get(':id')
  findOneEvent(@Param('id') id: string, @Query() params: ParamsFindSingleEvent) {
    params.id = id;
    return this.eventsService.findOne(params);
  }

  @Post('')
  createEvent(@Body() data: CreateEventDTO) {
    return this.eventsService.create(data);
  }

  @Put(':id')
  updateEvent(@Param('id') id: string, @Body() data: UpdateEventDTO) {
    return this.eventsService.update(id, data);
  }

  @Delete(':id')
  deleteEvent(@Param('id') id: string) {
    return this.eventsService.delete(id);
  }
}
