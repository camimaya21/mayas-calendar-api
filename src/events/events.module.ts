import { Module } from '@nestjs/common';
import { ConfigModule } from 'src/shared/config/config.module';
import { DatabaseModule } from 'src/shared/database/database.module';
import { EventService } from './events.service';

@Module({
  providers: [EventService],
  imports: [DatabaseModule, ConfigModule]
})
export class EventsModule {}
