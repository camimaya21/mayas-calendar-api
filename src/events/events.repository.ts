import { Injectable } from '@nestjs/common';
import { PrismaService } from '../shared/database/prisma.service';
import { Event } from '@prisma/client';
import {
  IEventsRepository,
  ParamsFindAllEvents,
  ParamsFindSingleEvent
} from './types/events.types';
import { CreateEventDTO, UpdateEventDTO } from './dto/events.dto';
import { toDate } from 'date-fns';

@Injectable()
export class EventsRepository implements IEventsRepository {
  constructor(private prisma: PrismaService) {}

  async findOne(params: ParamsFindSingleEvent): Promise<Event | null> {
    return this.prisma.event.findUnique({
      where: { id: params.id, internal_id: params.internalId }
    });
  }

  async findAll(params: ParamsFindAllEvents): Promise<Event[]> {
    const { cursor, where, orderBy } = params;
    return this.prisma.event.findMany({ cursor, where, orderBy });
  }

  async create(data: CreateEventDTO): Promise<Event> {
    return this.prisma.event.create({
      data: {
        title: data.title,
        description: data.description,
        start_date: toDate(new Date(data.startDate)),
        end_date: toDate(new Date(data.endDate))
      }
    });
  }

  async update(params: ParamsFindSingleEvent, data: UpdateEventDTO): Promise<Event> {
    return this.prisma.event.update({
      data: {
        title: data.title,
        description: data.description,
        start_date: toDate(new Date(data.startDate)),
        end_date: toDate(new Date(data.endDate))
      },
      where: { id: params.id, internal_id: params.internalId }
    });
  }

  async delete(params: ParamsFindSingleEvent): Promise<Event> {
    return this.prisma.event.delete({ where: { id: params.id, internal_id: params.internalId } });
  }
}
