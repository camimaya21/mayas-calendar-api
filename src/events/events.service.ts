import { Inject, Injectable, NotFoundException } from '@nestjs/common';
import { REPOSITORIES } from '../shared/constants/injectables.constants';
import { CreateEventDTO, DeleteEventResult, EventDTO, UpdateEventDTO } from './dto/events.dto';

import {
  IEventsRepository,
  IEventsService,
  ParamsFindAllEvents,
  ParamsFindSingleEvent
} from './types/events.types';

@Injectable()
export class EventsService implements IEventsService {
  constructor(
    @Inject(REPOSITORIES.IEventsRepository) private eventsRepository: IEventsRepository
  ) {}

  async findOne(params: ParamsFindSingleEvent): Promise<EventDTO | null> {
    const res = await this.eventsRepository.findOne({ id: params.id });
    return EventDTO.entityToDTO(res);
  }

  async findAll(params: ParamsFindAllEvents): Promise<EventDTO[]> {
    const res = await this.eventsRepository.findAll(params);
    return res.map(EventDTO.entityToDTO);
  }

  async create(data: CreateEventDTO): Promise<EventDTO> {
    const res = await this.eventsRepository.create(data);
    return EventDTO.entityToDTO(res);
  }

  async update(id: string, data: UpdateEventDTO): Promise<EventDTO> {
    await this._checkEventExists(id);
    const res = await this.eventsRepository.update({ id }, data);
    return EventDTO.entityToDTO(res);
  }

  async delete(id: string): Promise<DeleteEventResult> {
    await this._checkEventExists(id);
    const res = await this.eventsRepository.delete({ id });
    return { id: res.id, result: 'Event was deleted' };
  }

  private async _checkEventExists(id: string) {
    const event = await this.eventsRepository.findOne({ id });
    if (!event) {
      throw new NotFoundException(`Event with id ${id} not found`);
    }
    return event;
  }
}
