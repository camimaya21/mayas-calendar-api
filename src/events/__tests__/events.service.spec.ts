import { Test, TestingModule } from '@nestjs/testing';
import { TEST_EVENT_ID, EVENTS_REPOSITORY_MOCK_DATA } from '../__mocks__/events.mock.data';
import { REPOSITORIES, SERVICES } from '../../shared/constants/injectables.constants';
import { EventsService } from '../events.service';
import { CreateEventDTO, UpdateEventDTO } from '../dto/events.dto';
import { MockEventsRepository } from '../__mocks__/events.repository.mock';

describe('EventsService', () => {
  let service: EventsService;

  beforeEach(async () => {
    jest.clearAllMocks();

    const module: TestingModule = await Test.createTestingModule({
      providers: [
        EventsService,
        { provide: REPOSITORIES.IEventsRepository, useValue: MockEventsRepository }
      ]
    }).compile();

    service = module.get<EventsService>(EventsService);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  describe('Find methods', () => {
    it('should properly call events.repository.findAll', async () => {
      MockEventsRepository.findAll.mockResolvedValue(EVENTS_REPOSITORY_MOCK_DATA);
      await service.findAll({});
      expect(MockEventsRepository.findAll).toHaveBeenCalled();
    });

    it('should properly call events.repository.findOne', async () => {
      MockEventsRepository.findOne.mockResolvedValue(EVENTS_REPOSITORY_MOCK_DATA[0]);
      await service.findOne({ id: TEST_EVENT_ID });
      expect(MockEventsRepository.findOne).toHaveBeenCalledWith({ id: TEST_EVENT_ID });
    });
  });

  describe('Create methods', () => {
    it('should properly call events.repository.create', async () => {
      MockEventsRepository.findAll.mockResolvedValue([]);
      MockEventsRepository.create.mockResolvedValue(EVENTS_REPOSITORY_MOCK_DATA[0]);

      const eventData: CreateEventDTO = {
        title: 'some cool title',
        description: 'this is an example body',
        startDate: '2021-03-07T10:00:00.000Z',
        endDate: '2021-03-07T11:00:00.000Z'
      };

      const res = await service.create(eventData);

      expect(MockEventsRepository.create).toHaveBeenCalledWith(eventData);
      expect(res).toMatchObject({
        id: TEST_EVENT_ID,
        ...eventData
      });
    });
  });

  describe('Update methods', () => {
    const eventUpdateData: UpdateEventDTO = {
      title: 'A better title',
      description: 'this is an example body',
      startDate: '2021-03-07T10:00:00.000Z',
      endDate: '2021-03-07T11:00:00.000Z'
    };

    it('should properly call events.repository.update if event exist', async () => {
      MockEventsRepository.findOne.mockResolvedValue({ id: TEST_EVENT_ID });
      MockEventsRepository.update.mockResolvedValue(
        Object.assign(EVENTS_REPOSITORY_MOCK_DATA[0], {
          id: TEST_EVENT_ID,
          title: 'A better title'
        })
      );

      const res = await service.update(TEST_EVENT_ID, eventUpdateData);

      expect(MockEventsRepository.findOne).toHaveBeenCalledWith({ id: TEST_EVENT_ID });
      expect(MockEventsRepository.update).toHaveBeenCalledWith(
        { id: TEST_EVENT_ID },
        eventUpdateData
      );
      expect(res).toMatchSnapshot();
    });

    it('should fail if no event has found', async () => {
      MockEventsRepository.findOne.mockResolvedValue(undefined);

      await expect(() => service.update(TEST_EVENT_ID, eventUpdateData)).rejects.toThrowError(
        'Event with id 01123b18-cf8b-451f-a699-441397f02949 not found'
      );

      expect(MockEventsRepository.findOne).toHaveBeenCalledWith({ id: TEST_EVENT_ID });
      expect(MockEventsRepository.update).not.toHaveBeenCalled();
    });
  });
  describe('Delete methods', () => {
    it('should properly call events.repository.update if event exist', async () => {
      MockEventsRepository.findOne.mockResolvedValue(EVENTS_REPOSITORY_MOCK_DATA[0]);
      MockEventsRepository.delete.mockResolvedValue({ id: TEST_EVENT_ID });

      const res = await service.delete(TEST_EVENT_ID);

      expect(MockEventsRepository.findOne).toHaveBeenCalledWith({ id: TEST_EVENT_ID });
      expect(MockEventsRepository.delete).toHaveBeenCalledWith({ id: TEST_EVENT_ID });
      expect(res.id).toBe(TEST_EVENT_ID);
      expect(res.result).toBe('Event was deleted');
    });

    it('should fail if no event has found', async () => {
      MockEventsRepository.findOne.mockResolvedValue(undefined);

      await expect(() => service.delete(TEST_EVENT_ID)).rejects.toThrowError(
        'Event with id 01123b18-cf8b-451f-a699-441397f02949 not found'
      );

      expect(MockEventsRepository.findOne).toHaveBeenCalledWith({ id: TEST_EVENT_ID });
      expect(MockEventsRepository.delete).not.toHaveBeenCalled();
    });
  });
});
