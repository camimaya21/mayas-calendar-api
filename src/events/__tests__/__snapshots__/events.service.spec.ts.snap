// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventsService Update methods should properly call events.repository.update if event exist 1`] = `
Object {
  "description": "this is an example body",
  "endDate": "2021-03-07T11:00:00.000Z",
  "id": "01123b18-cf8b-451f-a699-441397f02949",
  "startDate": "2021-03-07T10:00:00.000Z",
  "title": "A better title",
}
`;
